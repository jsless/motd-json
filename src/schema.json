{
  "$id": "https://jsless.org/motd.schema.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Messages of the Day",
  "type": "array",
  "minItems": 1,
  "items": {
    "$ref": "#/definitions/motd"
  },
  "definitions": {
    "motd": {
      "type": "object",
      "required": [
        "message"
      ],
      "properties": {
        "message": {
          "type": "string",
          "description": "The message of the day",
          "minLength": 5
        },
        "period": {
          "$ref": "#/definitions/period"
        },
        "regions": {
          "type": "array",
          "minItems": 1,
          "items": {
            "$ref": "#/definitions/region"
          }
        },
        "tags": {
          "type": "array",
          "minItems": 1,
          "items": {
            "oneOf": [
              { "$ref": "#/definitions/booleanTag" },
              { "$ref": "#/definitions/numberTag" },
              { "$ref": "#/definitions/semverTag" },
              { "$ref": "#/definitions/setTag" }
            ]
          }
        }
      }
    },
    "period": {
      "type": "object",
      "required": [
        "from",
        "till"
      ],
      "properties": {
        "from": {
          "type": "string",
          "format": "date"
        },
        "till": {
          "type": "string",
          "format": "date"
        }
      }
    },
    "region": {
      "type": "string"
    },
    "semverTag": {
      "type": "object",
      "required": [
        "name",
        "type",
        "value"
      ],
      "properties": {
        "name": {
          "type": "string"
        },
        "type": {
          "enum": ["semver"]
        },
        "value": {
          "type": "string"
        }
      }
    },
    "numberTag": {
      "type": "object",
      "required": [
        "name",
        "type",
        "value"
      ],
      "properties": {
        "name": {
          "type": "string"
        },
        "type": {
          "enum": ["number"]
        },
        "value": {
          "type": "number"
        }
      }
    },
    "setTag": {
      "type": "object",
      "required": [
        "name",
        "type",
        "value"
      ],
      "properties": {
        "name": {
          "type": "string"
        },
        "type": {
          "enum": ["set"]
        },
        "value": {
          "type": "array"
        }
      }
    },
    "booleanTag": {
      "type": "object",
      "required": [
        "name",
        "type",
        "value"
      ],
      "properties": {
        "name": {
          "type": "string"
        },
        "type": {
          "enum": ["boolean"]
        },
        "value": {
          "type": "boolean"
        }
      }
    }
  }
}

