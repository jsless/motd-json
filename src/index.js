import osLocale from 'os-locale'
import semver from 'semver'
import schema from '../src/schema'

export function getRegion() {
  return osLocale.sync({ spawn: false })
    .substr(3, 2)
    .toLowerCase()
}

export function filterPeriod(data, now) {
  if (now && typeof now === 'string') {
    now = new Date(now)
  }

  now = now || new Date().getTime()
  const currentYear = `${new Date().getFullYear()}-`

  return data.filter(({ period: { from, till } = {} }) => {
    if (from && till) {
      from = new Date(from.replace('0000-', currentYear)).getTime()
      till = new Date(till.replace('0000-', currentYear)).getTime()

      return now >= from && now <= till
    }

    return true
  })
}

export function filterRegional(data, region) {
  if (region) {
    region = region.toLowerCase()
  } else {
    region = getRegion()
  }

  return data.filter(({ regions }) => !regions || regions.includes(region))
}

export function filterTags(data, tags) {
  if (!tags) {
    return data
  }

  return data.filter(({ tags: dataTags }) => {
    if (!dataTags) {
      return true
    }

    return dataTags.every(({ name, type, value }) => {
      if (!(name in tags)) {
        return true
      }

      const tagValue = tags[name]

      if (type === 'semver') {
        return semver.satisfies(tagValue, value)
      }

      if (type === 'number') {
        // TODO: make operator optionsurable
        return tagValue >= value
      }

      if (type === 'set') {
        if (Array.isArray(tagValue)) {
          return tagValue.some(aTagValue => value.includes(aTagValue))
        }
        return value.includes(tagValue)
      }

      if (type === 'boolean') {
        return value === tagValue
      }
    })
  })
}

export function filter(data, options) {
  const {
    region = '',
    regional = false,
    date = '',
    tags = false
  } = options || {}

  if (regional || region) {
    data = filterRegional(data, region)
  }

  if (date) {
    data = filterPeriod(data, date)
  }

  if (tags) {
    data = filterTags(data, tags)
  }

  return data
}

let ajvValidate

export async function validate(data) {
  if (!ajvValidate) {
    try {
      const Ajv = await import('ajv').then(m => m.default || m)
      ajvValidate = new Ajv().compile(schema)
    } catch (err) {
      if (err.code === 'MODULE_NOT_FOUND') {
        console.error('ajv not found, please install the peer dependency') // eslint-disable-line no-console
        return true
      }

      throw err
    }
  }

  return ajvValidate(data)
}

export function motd(data, options) {
  const motds = filter(data, options)

  const index = Math.floor(Math.random() * motds.length)

  const { message } = motds[index]
  return message || ''
}
