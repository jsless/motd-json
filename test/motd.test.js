import osLocale from 'os-locale'
import * as motd from '../src'

jest.mock('os-locale', () => ({
  sync: jest.fn()
}))

const data = [
  {
    message: 'my test message'
  },
  {
    message: 'my multi-region message',
    regions: ['en', 'nl']
  },
  {
    message: 'mijn test bericht',
    regions: ['nl']
  },
  {
    message: 'Merry Christmas',
    period: {
      from: '0000-12-01',
      till: '0000-12-27'
    }
  },
  {
    message: 'Fijne Sinterklaas',
    regions: ['nl'],
    period: {
      from: '2019-11-16',
      till: '2019-12-05'
    }
  }
]

describe('motd', () => {
  beforeEach(() => {
    osLocale.sync.mockReturnValue('nl-NL')
  })

  afterEach(() => {
    jest.resetAllMocks()
    jest.restoreAllMocks()
  })

  test('schema', async () => {
    await expect(motd.validate(data)).resolves.toBe(true)
  })

  test('getRegion only returns region part from locale', () => {
    expect(motd.getRegion()).toBe('nl')

    osLocale.sync.mockReturnValue('en_UK')
    expect(motd.getRegion()).toBe('uk')
  })

  test('filterPeriod returns motds for current period only', () => {
    let now = new Date('2019-11-20')
    expect(motd.filterPeriod(data, now)).toMatchSnapshot()

    now = new Date('2019-12-01')
    expect(motd.filterPeriod(data, now)).toMatchSnapshot()

    now = new Date('2019-12-10')
    expect(motd.filterPeriod(data, now)).toMatchSnapshot()
  })

  test('filterRegional only returns motds for current region', () => {
    expect(motd.filterRegional(data)).toMatchSnapshot()
    expect(motd.filterRegional(data, 'en')).toMatchSnapshot()
    expect(motd.filterRegional(data, 'fr')).toMatchSnapshot()
  })

  test('filterMotds returns all', () => {
    expect(motd.filter(data, { date: '2019-12-01' })).toMatchSnapshot()
  })

  test('motd returns message', () => {
    const options = {
      region: 'fr',
      date: '2019-01-01',
      tags: {}
    }

    expect(motd.filter(data, options)).toMatchSnapshot()
    expect(motd.motd(data, options)).toBe('my test message')
  })
})
