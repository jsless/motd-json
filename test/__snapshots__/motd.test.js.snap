// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`motd filterMotds returns all 1`] = `
Array [
  Object {
    "message": "my test message",
  },
  Object {
    "message": "my multi-region message",
    "regions": Array [
      "en",
      "nl",
    ],
  },
  Object {
    "message": "mijn test bericht",
    "regions": Array [
      "nl",
    ],
  },
  Object {
    "message": "Merry Christmas",
    "period": Object {
      "from": "0000-12-01",
      "till": "0000-12-27",
    },
  },
  Object {
    "message": "Fijne Sinterklaas",
    "period": Object {
      "from": "2019-11-16",
      "till": "2019-12-05",
    },
    "regions": Array [
      "nl",
    ],
  },
]
`;

exports[`motd filterPeriod returns motds for current period only (2019-11-20) 1`] = `
Array [
  Object {
    "message": "my test message",
  },
  Object {
    "message": "my multi-region message",
    "regions": Array [
      "en",
      "nl",
    ],
  },
  Object {
    "message": "mijn test bericht",
    "regions": Array [
      "nl",
    ],
  },
  Object {
    "message": "Fijne Sinterklaas",
    "period": Object {
      "from": "2019-11-16",
      "till": "2019-12-05",
    },
    "regions": Array [
      "nl",
    ],
  },
]
`;

exports[`motd filterPeriod returns motds for current period only (2019-12-01) 1`] = `
Array [
  Object {
    "message": "my test message",
  },
  Object {
    "message": "my multi-region message",
    "regions": Array [
      "en",
      "nl",
    ],
  },
  Object {
    "message": "mijn test bericht",
    "regions": Array [
      "nl",
    ],
  },
  Object {
    "message": "Merry Christmas",
    "period": Object {
      "from": "0000-12-01",
      "till": "0000-12-27",
    },
  },
  Object {
    "message": "Fijne Sinterklaas",
    "period": Object {
      "from": "2019-11-16",
      "till": "2019-12-05",
    },
    "regions": Array [
      "nl",
    ],
  },
]
`;

exports[`motd filterPeriod returns motds for current period only (2019-12-10) 1`] = `
Array [
  Object {
    "message": "my test message",
  },
  Object {
    "message": "my multi-region message",
    "regions": Array [
      "en",
      "nl",
    ],
  },
  Object {
    "message": "mijn test bericht",
    "regions": Array [
      "nl",
    ],
  },
  Object {
    "message": "Merry Christmas",
    "period": Object {
      "from": "0000-12-01",
      "till": "0000-12-27",
    },
  },
]
`;

exports[`motd filterPeriod returns motds for current period only (2020-01-04) 1`] = `
Array [
  Object {
    "message": "my test message",
  },
  Object {
    "message": "my multi-region message",
    "regions": Array [
      "en",
      "nl",
    ],
  },
  Object {
    "message": "mijn test bericht",
    "regions": Array [
      "nl",
    ],
  },
]
`;

exports[`motd filterRegional only returns motds for current region 1`] = `
Array [
  Object {
    "message": "my test message",
  },
  Object {
    "message": "my multi-region message",
    "regions": Array [
      "en",
      "nl",
    ],
  },
  Object {
    "message": "mijn test bericht",
    "regions": Array [
      "nl",
    ],
  },
  Object {
    "message": "Happy New Year",
    "periods": Array [
      Object {
        "from": "0000-12-28",
        "till": "0000-12-31",
      },
      Object {
        "from": "0000-01-01",
        "till": "0000-01-05",
      },
    ],
  },
  Object {
    "message": "Merry Christmas",
    "period": Object {
      "from": "0000-12-01",
      "till": "0000-12-27",
    },
  },
  Object {
    "message": "Fijne Sinterklaas",
    "period": Object {
      "from": "2019-11-16",
      "till": "2019-12-05",
    },
    "regions": Array [
      "nl",
    ],
  },
]
`;

exports[`motd filterRegional only returns motds for current region 2`] = `
Array [
  Object {
    "message": "my test message",
  },
  Object {
    "message": "my multi-region message",
    "regions": Array [
      "en",
      "nl",
    ],
  },
  Object {
    "message": "Happy New Year",
    "periods": Array [
      Object {
        "from": "0000-12-28",
        "till": "0000-12-31",
      },
      Object {
        "from": "0000-01-01",
        "till": "0000-01-05",
      },
    ],
  },
  Object {
    "message": "Merry Christmas",
    "period": Object {
      "from": "0000-12-01",
      "till": "0000-12-27",
    },
  },
]
`;

exports[`motd filterRegional only returns motds for current region 3`] = `
Array [
  Object {
    "message": "my test message",
  },
  Object {
    "message": "Happy New Year",
    "periods": Array [
      Object {
        "from": "0000-12-28",
        "till": "0000-12-31",
      },
      Object {
        "from": "0000-01-01",
        "till": "0000-01-05",
      },
    ],
  },
  Object {
    "message": "Merry Christmas",
    "period": Object {
      "from": "0000-12-01",
      "till": "0000-12-27",
    },
  },
]
`;

exports[`motd motd returns message 1`] = `
Array [
  Object {
    "message": "my test message",
  },
]
`;
